### ISSUE TITLE
WrapperTask triggers many unnecessary complete() calls

### ISSUE THREAD

**Issue by mila** [spotify/luigi#2994]
## Description

To determine completeness of `WrapperTask`, the `complete()` method has to be called on its dependencies.
Because a scheduler is not aware of this check, it will call the `complete()` method on the dependencies for the second time.
And when `check_unfulfilled_deps` is enabled (default), Luigi calls the `complete()` method on all dependencies for the third time.

When a wrapper task with 99 complete dependencies and 1 incomplete, this can result in 300 completeness checks.

This problem is multiplied if we build a hierarchy of wrapper task because all `WrapperTask.complete()` calls have to bubble to leaf tasks. For example consider a hierarchy of three wrapper tasks:

```python

import luigi
import luigi.mock

class Dep(luigi.Task):
    def output(self): return luigi.mock.MockTarget("inner.txt")
    def run(self): self.output()._touchz()
    def complete(self): print("complete?"); return super().complete()

class BottomWrapper(luigi.WrapperTask):
    def requires(self): return Dep()

class MiddleWrapper(luigi.WrapperTask):
    def requires(self): return BottomWrapper()

class TopWrapper(luigi.WrapperTask):
    def requires(self): return MiddleWrapper()
```
Running the `TopWrapper` from the example results in 7 calls to `Dep.complete()`.

## Proposed solution

From my experiments, it seems that I can replace Luigi `WrapperTask` by the following `PhonyTask`:

```python
class PhonyTask(luigi.Task):

    _complete = False

    def run(self):
        self._complete = True

    def complete(self):
        return self._complete
```
The difference is that scheduler reports the task as "ran successfully" instead of "complete one", but I don't see any problem with it.

We can add `PhonyTask` to Luigi, or we can reimplement `WrapperTask` using the above fashion.

Alternatively, Luigi could handle wrapper tasks as a special case, looking directly at their dependencies and skipping its complete() method.

**Comment by mila** [comment]
Any thoughts about this issue? I can create a PR with the `PhonyTask` class.

**Comment by freider** [comment]
I agree these repeated checks are far from optimal, especially in the nested wrapper scenario you show.
The reason it's there to begin with is that a task is both checked for completeness when the dependency tree is first expanded, and then every input task is checked before a dependent task starts (if check_unfulfilled_deps) to make sure that:
1. the task running without exceptions actually resulted in complete() being true
2. the completeness hasn't been invalidated since the task ran
The intention with the above is to prevent to some best effort to never run a subsequent task if a dependency hasn't run.

I've been thinking about a good way of solving this lately, for tasks and wrappers where it might not be what you want. Perhaps letting a task define itself as "safe" or "atomic" to signify that if it ran successfully it guarantees that complete() is true immediately afterwards. This would be the same as setting check_unfulfilled_deps=False, but on a per-dependency level, which allows for better control. In addition to that, what you might want is a centralized timed cache for  true `complete()` results, also configurable on a task-basis.

The problem with PhonyTask in your fix is that it will only work for a single luigi host/process, since the result is cached in the in-memory task instance. You can also never re-invalidate the task, but that might be less of a problem when only running short-lived workers on a single machine. Adding that logic only to the wrapper classes will still not fix the problem of double-checking a task both during DAG expansion and before the PhonyTask is granted run permission, but it will reduce the explosion of checks caused by nested wrapper tasks (again, only when running a single process).

I hope I'll have time to contribute something in this area in the near future :)

**Comment by mila** [comment]
Thank you @freider for your explanation. I am aware that I proposed a very radical (and backward incompatible) change.
I have not enough insight to anticipate all possible problems, so your response is very helpful for me. 

I see the proposed `PhonyTask` as an alternative to the builtin `WrapperTask`. The difference is that `PhonyTask` is never complete, it is always executed again (similar to phony targets in Make). My idea is that it is cheaper to execute an empty run() method than to check whether dependencies are complete.

You correctly noticed that I set `self._complete = True` only because of `check_unfulfilled_deps`, and you are right that this works only in a single luigi host/process. But when I use `check_complete_on_run=true` and `check_unfulfilled_deps=false` (see #2992), I may not notice the issue because only one worker will check the status.

A systematic solution would be to add first-class support for phony tasks. For example, a phony task would not define `complete()` (similar to the missing `ExternalTask.run()`). The scheduler would always run such tasks.

Marking tasks as "safe" (setting `check_unfulfilled_deps=False` per task) would save the last extra check, but the first one would remain. If you implemented safe tasks, I would wrote my `PhonyTask` as a safe task with `def complete(self): return False`.

**Comment by lallea** [comment]
@freider We have use cases where I feel that a shared target state cache or something similar would make a significant difference. We often have large dependency trees where most tasks are complete, but sometimes there is a large backfill. I have a hunch that a database as a cache for GCS/S3/HDFS storage metadata would work well. 

Ping me if you would like to sync on needs and design. Good to see you back active with Luigi again, BTW. :-)
